applications/ParticleMechanicsApplication/custom_constitutive/johnson_cook_thermal_plastic_3D_law.cpp

```diff
@@ -0,0 +1,476 @@
+				if (MaterialProperties[TAYLOR_QUINNEY_COEFFICIENT] > 0.0) {
+					dYield_dGamma += MaterialProperties[TAYLOR_QUINNEY_COEFFICIENT] * yield_stress/ MaterialProperties[DENSITY] /
+						MaterialProperties[SPECIFIC_HEAT] * CalculateThermalDerivative(MaterialProperties, predicted_eps, predicted_eps_rate, predicted_temperature);
+				}
+				dYield_dGamma *= GetSqrt23();
+				yield_function_gradient = -1.0 * GetSqrt6() * shear_modulus_G - dYield_dGamma;
+
+				// Update gamma and check for convergence
+				delta_gamma = -1.0 * yield_function / yield_function_gradient;
+				if (std::abs(delta_gamma) < tolerance_delta_gamma) {
+					is_converged = true;
+					break;
+				}
+				else gamma += delta_gamma;
+
+				// Bisect increment if out of search bounds
+				if (gamma_min > gamma || gamma > gamma_max) gamma = gamma_min + 0.5 * (gamma_max - gamma_min);
+
+				// Update of quantities
+				predicted_eps = mEquivalentPlasticStrainOld + GetSqrt23() * gamma; // eps = equivalent plastic strain
+				predicted_eps_rate = GetSqrt23() * gamma / CurrentProcessInfo[DELTA_TIME];
+				predicted_temperature = mTemperatureOld;
+				if (MaterialProperties[TAYLOR_QUINNEY_COEFFICIENT] > 0.0) {
+					predicted_temperature += MaterialProperties[TAYLOR_QUINNEY_COEFFICIENT] / GetSqrt6() / MaterialProperties[DENSITY]
+						/ MaterialProperties[SPECIFIC_HEAT] * (yield_stress + mYieldStressOld) * gamma;
+				}
+
+				iteration += 1;
+				if (iteration == iteration_limit)
+				{
+					KRATOS_INFO("Johnson Cook Material Model") << " Johnson Cook iteration limit exceeded\n";
+					KRATOS_WATCH(gamma)
+					KRATOS_WATCH(delta_gamma)
+					KRATOS_WATCH(yield_function)
+					KRATOS_ERROR << "Johnson Cook iteration limit exceeded";
+				}
+			}
+			// Correct trial stress
+			stress_deviatoric_converged = stress_deviatoric_trial - 2.0 * shear_modulus_G * gamma * flow_direction_normalized;
+
+			// Store plastic deformation quantities
+			mEnergyDissipated += gamma / GetSqrt6() / MaterialProperties[DENSITY] * (mYieldStressOld + yield_stress);
+			mEquivalentPlasticStrainOld = predicted_eps;
+			mPlasticStrainRateOld = predicted_eps_rate;
+			mTemperatureOld = predicted_temperature;
+			mYieldStressOld = yield_stress;
+			mHardeningRatio = yield_stress / mYieldStressVirgin;
+		}
+		else
+		{
+			if (yield_stress / mYieldStressVirgin > yield_stress_failure_ratio)
+			{
+				stress_deviatoric_converged = stress_deviatoric_trial;
+			}
+			else
+			{
+				// Particle has failed. It can only take compressive volumetric stresses!
+				stress_deviatoric_converged.clear();
+				if (stress_hydrostatic_new > 0.0) stress_hydrostatic_new = 0.0;
+				mHardeningRatio = 0.0;
+			}
+		}
+		// Update equivalent stress
+		Matrix stress_converged = stress_deviatoric_converged + stress_hydrostatic_new * identity;
+		mEquivalentStress = std::sqrt(3.0 / 2.0 * CalculateMatrixDoubleContraction(stress_deviatoric_converged));
```

Comment:
Maybe this should be implemented in the stress utility too?

Reply:
Yep that's a good idea 👍 - I did it for the double contraction and the trace function